dxr/plugins/python/analysis.py

```diff
@@ -202,4 +203,70 @@ def analyze_import(self, node):
                     if package_path:
                         import_name = package_path + '.' + import_name
 
-            self.tree_analysis.names[absolute_local_name] = import_name
+            import_tuple = Import(absolute_name=absolute_local_name,
+                                  line=node.lineno, col=node.col_offset,
+                                  import_name=import_name)
+            self.tree_analysis.definition_tree.add(import_tuple)
+
+
+# Namedtuples that are stored in DefinitionTree.
+Definition = namedtuple('Definition', ('absolute_name', 'line', 'col', 'path'))
+Import = namedtuple('Import', ('absolute_name', 'line', 'col', 'import_name'))
+
+
+class DefinitionTree(object):
+    """Stores definitions across the entire project, such as classes and
+    functions, as well as the imports that link them together.
+
+    """
+    def __init__(self):
+        self.modules = {}
+
+    def add(self, definition):
+        """Add a definition to the tree. """
+        names = definition.absolute_name.split('.')
+        scope = self.modules
+
+        for name in names[:-1]:
+            scope = scope.setdefault(name, {})
+        scope[names[-1]] = definition
+
+    def get(self, name):
+        """Fetch the definition for the given name, following imports if
+        necessary.
+
+        If the definition is not found, return None. Top-level modules
+        that have not been added to the tree are assumed external and
+        valid, and definitions will be returned for them with no path
+        or position data.
+        """
+        names = name.split('.')
+        module_name = names.pop(0)
```

Comment:
I'm still trying to work out whether this is recursive, but keep in mind that popping things off the left of a list is inefficient: it has to memcpy all the other items in the list down a notch.